src/Crypto/WebAuthn/Model/Types.hs

```diff
@@ -7,11 +7,40 @@
 {-# LANGUAGE TypeFamilies #-}
 
 -- | Stability: experimental
--- This module contains the same top-level definitions as 'Crypto.WebAuthn.Client.JavaScript',
--- but with the types containing a more Haskell-friendly structure.
+-- This module contains Haskell-friendly types for structures used in WebAuthn
+-- that are used throughout this library. These types are modelled according to
+-- the following conventions:
 --
--- Note: The 'ToJSON' instances of these types are for pretty-printing purposes
--- only.
+-- * If a structure has the same semantics for both the
+--   [registration](https://www.w3.org/TR/webauthn-2/#registration) and
+--   [authentication](https://www.w3.org/TR/webauthn-2/#authentication) WebAuthn
+--   [ceremonies](https://www.w3.org/TR/webauthn-2/#ceremony), then its type is
+--   parametrized by a @c@ parameter of kind 'CeremonyKind'. If such types have
+--   differing fields,
+--   [GADTs](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/gadt.html)
+--   are used to distinguish between them, where the constructor name is the
+--   type name with a @...Registration@ or @...Authentication@ suffix
+-- * If the raw bytes are needed for verification purposes of a structure, then
+--   its type is parametrized by a @raw@ parameter of kind 'Bool'. Only if @raw
+--   ~ 'True'@, the raw bytes of the necessary structures has to be present in
+--   the type. The type 'RawField' is used as a helper type for this.
+-- * In order to avoid duplicate record fields, all fields are prefixed with
+--   the initials of the constructor name.
+-- * Every type should have a 'ToJSON' instance for pretty-printing purposes.
+--   This JSON encoding doesn't correspond to any encoding used for
+--   sending/receiving these structures, it's only used for pretty-printing,
+--   which is why it doesn't need to be standardized. For encoding these
+--   structures from/to JSON for sending/receiving, see the
+--   'Crypto.WebAuthn.Encoding.WebAuthnJson' module
+-- * Fields of the WebAuthn standard that are optional (for writing) but have
+--   defaults (making them non-optional for reading) are encoded as
+--   non-optional fields, while the defaults are exposed in the
+--   'Crypto.WebAuthn.Model.Defaults' module. The alternative of making these
+--   fields optional would allow RP not having to specify them, which seems
+--   like a less safer option, since the defaults might not be what is really
+--   needed, and they might change. The root cause why this decision had to be
+--   made is that such assymetrical reading/writing fields don't map nicely to
+--   Haskell's records.
```

Comment:
After discussion with @ErinvanderVeen and some more thinking myself, I decided to not change the model types to optional ones, with the rationale given here. This makes the JSON encoding/decoding a bit less nice, but it's not too bad